arch-vega: Add pagetable walker buffer

TomXia opened the pull request:
Add a 64-entry PTE buffer to the pagetable walker. It significantly reduces memory accesses, improving address translation performance.

TomXia committed:
arch-vega: Add pagetable walker buffer

Add a 64-entry PTE buffer to the pagetable walker. It
significantly reduces memory accesses, improving address
translation performance.

mattsinc commented on src/arch/amdgpu/vega/pagetable_walker.cc:
```diff
@@ -101,6 +101,45 @@ Walker::WalkerState::startFunctional(Addr base, Addr vaddr,
+{
+    auto entry = entryList.begin();
+
+    for ( ; entry != entryList.end(); ++entry)
```
please add a comment explaining what this is doing

powerjg replied:
```suggestion
    for (auto entry : entryList) {
```
Something like that would be better and more standard modern C++

mattsinc replied:
@TomXia : still need to make change @powerjg requested here

TomXia replied:
Thanks for pointing out.
To track entries in LRU order, the container might be modified. Range based for might not be a direct way to implement it. Shall I still proceed to change?

mattsinc replied:
How does the change impact LRU order?

TomXia replied:
entryList (now renamed to pwcEntryList) here maintains the LRU-ordered list of current cached items. Thus, the list needs to be updated in this lookup method. Using regular for loop, it is simper to reorder the list within the loop because the iterators automatically come with it.

Although, there might be a different and efficient way to achieve it with range-based for that I am not aware of.

mattsinc replied:
Maybe I'm just missing something, but I'm asking for why this:

for ( ; entry != entryList.end(); ++entry)

and this:

for (auto entry : entryList) {

would lead to a different update order.  To me, they seem like they would both result in the same order ... I'm pretty sure the second option is the same as the first but using newer C++ primitives?

powerjg replied:
Here's the problem with this code: The comment isn't correct. It's not "searching the buffer for the paddr." It's updating the LRU position of an entry. If the code did what the comment said, then it would be read-only and a range-based for loop would work. What you're actually doing is manipulating the list inside an iteration (which is error prone, but looks OK in this case). As you said @TomXia you probably can't use the range-based for given you're changing the list.

(I will also go back to my first comment which is if you used the `AssociativeCache` from base/ you wouldn't have to worry about any of this.)

mattsinc commented on src/arch/amdgpu/vega/pagetable_walker.cc:
```diff
@@ -165,8 +204,8 @@ Walker::WalkerState::startWalk()
 
-        sendPackets();
         started = true;
+        sendPackets();
```
why did you invert the order of these two lines?

abmerop replied:
You can drop this change. Generally want to touch as few lines of code as possible

TomXia replied:
It must be reordered here. Because, sendPackets() might immediately have the response due to a buffer hit, which causes the WalkerState to change before sendPackets() returns. Therefore, the state (started) needs to be updated before sendPackets().

mattsinc commented on src/arch/amdgpu/vega/pagetable_walker.cc:
```diff
@@ -366,23 +405,36 @@ Walker::WalkerState::sendPackets()
-        DPRINTF(GPUPTWalker, "Timing request for %#lx successful\n",
-                read->getAddr());
     }
+    //No lines after sendTiming because the state might be freed
```
if this isn't used anymore then we should remove these lines instead of commenting them out

abmerop replied:
I am inclined to keep this code. It's something I look for when debugging translation issues so you can see which level of the page table you received a "weird" page table entry

mattsinc commented on src/arch/amdgpu/vega/pagetable_walker.cc:
```diff
@@ -427,8 +492,9 @@ void
-    for (iter = currStates.begin(); iter != currStates.end(); iter++) {
+    for (iter = currStates.begin(); iter != currStates.end(); ) {
         WalkerState * walkerState = *(iter);
+        iter++;
```
why are you incrementing iter here vs. at the bottom of the loop?  Also seems a little weird you had to move it at all...

abmerop replied:
Agree, move this back

TomXia replied:
Similarly, after retry(), the current iterator might be erased, because the request is immediately returned and finishes the walk. We need to move to the next iterator before the current one becomes invalid.

mattsinc replied:
I'm not sure why moving it here accomplishes that goal?

mattsinc commented on src/arch/amdgpu/vega/pagetable_walker.hh:
```diff
@@ -54,6 +54,11 @@ namespace VegaISA
 class Walker : public ClockedObject
 {
   protected:
+  typedef std::pair<Addr, PageTableEntry> paddr_pte_t;
```
please add comments explaining what each of these are used for

mattsinc replied:
this remains

mattsinc commented on src/arch/amdgpu/vega/pagetable_walker.hh:
```diff
@@ -199,6 +215,12 @@ class Walker : public ClockedObject
     {
         DPRINTF(GPUPTWalker, "Walker::Walker %p\n", this);
+        pte_buffer = false;
+        int num_sets = 64;
```
probably we want to make this a configurable thing eventually, but probably fine for now

abmerop replied:
I would also make this configurable immediately. You can add a parameter here https://github.com/gem5/gem5/blob/develop/src/arch/amdgpu/vega/VegaGPUTLB.py#L38 e.g., page_walk_cache_size and then replace num_sets with p.page_walk_cache_size

mattsinc commented on src/arch/amdgpu/vega/tlb.cc:
```diff
@@ -84,6 +84,7 @@ GpuTLB::GpuTLB(const VegaGPUTLBParams &p)
     missLatency1 = p.missLatency1;
     missLatency2 = p.missLatency2;
 
+
```
remove this

mattsinc replied:
this remains

mattsinc commented on src/arch/amdgpu/vega/tlb.cc:
```diff
@@ -104,6 +105,8 @@ GpuTLB::GpuTLB(const VegaGPUTLBParams &p)
         gpuDevice->getVM().registerTLB(this);
     }
+
+    walker->setEnablePTEBuffer(true);
```
probably we want the walker to be configurable eventually, but ok for now.

abmerop replied:
This should be moved into the walker itself.  You could technically not have TLBs, then the this feature would not be used. I would configure this right away you can add a parameter here with the default value of true: https://github.com/gem5/gem5/blob/develop/src/arch/amdgpu/vega/VegaGPUTLB.py#L38 .

mattsinc reviewed (changes requested):
I'm not sure the commit message is right -- I think what you've done here is add a page walk cache (PWC) to the page table walker (PTW).  We should probably update the commit message to be very explicit about this.

In general this looks good though, thanks!

powerjg reviewed (commented):
If you're adding a new cache to gem5, you really should use the generic cache objects so that you can use replacement policies, tagging policies, etc.

See https://github.com/gem5/gem5/tree/stable/src/base/cache and how that code is used all over the codebase.

abmerop reviewed (changes requested):
I marked few changes that shouldn't take too long or need retesting.

Optionally I would recommend some more descriptive variable/function names for things like buffer, insert, lookup that make it clear they are for the page walk cache and avoid confusion that you are looking up/inserting to the page table itself

mattsinc commented:
@TomXia how are the changes coming along that @abmerop and I requested?

@powerjg : we had discussed using that instead, and decided against it because this isn't a true cache (which is what @TomXia referred to is as a buffer instead).  @abmerop do you remember why we opted not to use that cache format?

abmerop commented:
> @powerjg : we had discussed using that instead, and decided against it because this isn't a true cache (which is what @TomXia referred to is as a buffer instead). @abmerop do you remember why we opted not to use that cache format?

This is so small it would not be implemented as a cache in real hardware

mattsinc commented on src/arch/amdgpu/vega/pagetable_walker.cc:
```diff
@@ -366,23 +405,36 @@ Walker::WalkerState::sendPackets()
     if (retrying)
         return;
-
+    auto addr = read->getAddr();
```
Since this variable is only used a DPRINTF for LLVM compilation I think we need to use the __maybe__unused__ declaration?

mattsinc replied:
I think this still needs to be done?

mattsinc reviewed (changes requested):
Getting closer!  Some questions, as well as some changes that still are needed

TomXia committed:
arch-vega: Add page walk cache (PWC)

Add a small cache buffering PTEs to the pagetable walker.
It significantly reduces memory accesses, improving address
translation performance.

TomXia commented:
Thanks for the comments. A new commit has been pushed, with a new commit message.

mattsinc commented on src/arch/amdgpu/vega/pagetable_walker.cc:
```diff
@@ -101,6 +101,44 @@ Walker::WalkerState::startFunctional(Addr base, Addr vaddr,
+    *newEntry = std::make_pair(paddr, entry);
+    pwcEntryList.push_front(newEntry);
+}
+PageTableEntry* Walker::pwcLookup(Addr paddr)
```
There should be a line of space between these functions

mattsinc commented on src/arch/amdgpu/vega/pagetable_walker.cc:
```diff
@@ -427,8 +491,9 @@ void
-    for (iter = currStates.begin(); iter != currStates.end(); iter++) {
+    for (iter = currStates.begin(); iter != currStates.end(); ) {
         WalkerState * walkerState = *(iter);
+        iter++;//increment before it becomes invalid
```
Please put a space between semicolon and comment

mattsinc replied:
Also probably need a more descriptive comment -- you need to explain that the retry() function may update the current iterator element, hence updating iter ahead of time is needed (or something like this)

mattsinc commented on src/arch/amdgpu/vega/VegaGPUTLB.py:
```diff
@@ -41,6 +41,8 @@ class VegaPagetableWalker(ClockedObject):
     cxx_header = "arch/amdgpu/vega/pagetable_walker.hh"
     port = RequestPort("Port for the hardware table walker")
     system = Param.System(Parent.any, "system object")
+    b_size = Param.Int(64, "Page walk cache size")
```
I'm not sure b_size is a sufficiently descriptive name so that it's clear.  I would prefer pwcSize or something like that

mattsinc replied:
likewise @abmerop recommended page_walk_cache_size

mattsinc commented on src/arch/amdgpu/vega/pagetable_walker.hh:
```diff
@@ -199,6 +214,14 @@ class Walker : public ClockedObject
         deviceRequestorId(999), system(p.system)
     {
         DPRINTF(GPUPTWalker, "Walker::Walker %p\n", this);
+        pte_buffer = p.enable_pte_buffer;
```
probably we want a more descriptive name than pte_buffer -- pte_buffer sounds more like the name of a pointer to the pteBuffer.

mattsinc reviewed (changes requested):
Getting closer!

TomXia committed:
arch-vega: Add page walk cache (PWC)

Add a small cache buffering PTEs to the pagetable walker.
It significantly reduces memory accesses, improving address
translation performance.

TomXia commented:
New commit pushed.

powerjg commented:
> > @powerjg : we had discussed using that instead, and decided against it because this isn't a true cache (which is what @TomXia referred to is as a buffer instead). @abmerop do you remember why we opted not to use that cache format?
> 
> This is so small it would not be implemented as a cache in real hardware

That's fine. I'm happy for this to be merged without this change.

*However*, even if this isn't a cache in hardware, that doesn't mean that the `AssociativeCache` (found in `src/base/cache`) class isn't the right way to model it in gem5. There's a couple of benefits to using the `AssociativeCache`. (1) Right now, this is LRU, but what if you wanted to change that to random or NMRU or anything else? If you use `AssociativeCache` all of the replacement policies are implemented *for you*. *You have no work to do!* (2) Maybe in the future, someone wants to investigate a more complex structure (e.g., not fully associative). Again, if you use `AssociativeCache` then there's literally no code that needs to be changed to do that. You get the features for free.

The downside is low (very little complexity is added to the code), but the potential upside is high. It was just a suggestion, so no need to take action, but I would encourage you to consider these tradeoffs when submitting future gem5 changes.

mattsinc reviewed (changes requested):
I think the main thing that remains is the auto change @powerjg asked for, which I don't totally follow.  It looks like everything else is resolved!

powerjg commented on src/arch/amdgpu/vega/pagetable_walker.hh:
```diff
@@ -199,6 +214,14 @@ class Walker : public ClockedObject
+        pwcBuffer.assign(buf_size, std::make_pair((Addr)0, (PageTableEntry)0));
+        for (int buf_i = 0; buf_i < buf_size; ++buf_i) {
+            pwcFreeList.push_back(&pwcBuffer.at(buf_i));
+        if (enable_pwc)
+          DPRINTF(GPUPTWalker, "PWC enabled, size %d\n", pwcBuffer.size());
```
```suggestion
          if (enable_pwc) {
               DPRINTF(GPUPTWalker, "PWC enabled, size %d\n", pwcBuffer.size());
          }
```
Indentation is wrong too

powerjg commented on src/arch/amdgpu/vega/pagetable_walker.hh:
```diff
@@ -54,6 +54,15 @@ namespace VegaISA
 class Walker : public ClockedObject
 {
   protected:
+  //PTE indexed by paddr
+  typedef std::pair<Addr, PageTableEntry> paddr_pte_t;
+  //the PTE buffer
+  std::vector<paddr_pte_t> pwcBuffer;
+  //List of valid entries in the buffer, LRU ordered
+  std::list<paddr_pte_t*> pwcEntryList;
+  //List of free entries in the buffer
+  std::list<paddr_pte_t*> pwcFreeList;
+
```
Indentation is wrong

powerjg commented on src/arch/amdgpu/vega/pagetable_walker.cc:
```diff
@@ -412,11 +464,24 @@ Walker::recvTimingResp(PacketPtr pkt)
 
     DPRINTF(GPUPTWalker, "Got response for %#lx from walker %p -- %#lx\n",
             pkt->getAddr(), senderState->senderWalk, pkt->getLE<uint64_t>());
+    if (enable_pwc && pwcLookup(pkt->getAddr()) == nullptr) //not found, push
+        pwcInsert(pkt->getAddr(), pkt->getLE<uint64_t>());
```
Please use curly braces... I know it seems silly, but until you spend hours tracking down a bug because someone else thought they would save themselves 2 characters...

powerjg commented on src/arch/amdgpu/vega/pagetable_walker.cc:
```diff
@@ -366,23 +405,36 @@ Walker::WalkerState::sendPackets()
     pkt->pushSenderState(walker_state);
 
+    PageTableEntry *entry = pwcLookup(pkt->getAddr());
+    if (entry != nullptr)
+    {
```
Style
```suggestion
    if (entry != nullptr) {
```

powerjg reviewed (commented):
There are many style issues still.

mattsinc commented:
@TomXia just checking if you are working on addressing @powerjg 's feedback?

TomXia committed:
arch-vega: Add page walk cache (PWC)

Add a small cache buffering PTEs to the pagetable walker.
It significantly reduces memory accesses, improving address
translation performance.

TomXia committed:
arch-vega: Add page walk cache (PWC)

Add a small cache buffering PTEs to the pagetable walker.
It significantly reduces memory accesses, improving address
translation performance.

TomXia commented:
Thanks @powerjg . I have rewrite with `AssociativeCache`. Also fixed the styling issues. I have taken notes on styling rules and will ensure they are strictly followed.

powerjg reviewed (commented):
First of all, thanks for taking the time to use the base indexing policy and more generic cache. There's a lot less code in the change now!

Looking at this implementation, I got to thinking... The base cache/indexing policy doesn't really work correctly for what you're trying to do. A better approach would be to create your own indexing policy instead of using the set associative one.

This isn't super straightforward, so I took a crack at it for you. See my changes here: https://github.com/powerjg/gem5/tree/pr/TomXia/2162. Specifically https://github.com/gem5/gem5/commit/c08e54da988479f8b7bd721d18d1f4c48301a22a.

Feel free to take that code and squash it all into a single commit authored by you.

However, as I was re-writing the code, I had a question about the intention/implementation. Was the intention to cache *only a single PTE* in an entry in the page walk cache? I.e., is this comment correct? https://github.com/gem5/gem5/commit/cd1c7fc7c867524eb50d31ea059575c180855ff5#diff-e45ce01b5e603be741b8372fbfe1969965c7cb939464c5910cfba1e123f64f99R80

If that was the intention, then how does this structure differ from the TLB?

Finally, **I did not test this code**. The system I'm on doesn't have an easy way for me to test it. It compiles, but if you think this is a good way forward then make sure you test it :).

The main problem I had with the current implementation was the way the indexing policy was hard coded in the SimObject description file. It is bad practice to *instantiate* a SimObject in the "SimObject description" file. This file is for specifying parameters (and some defaults). 
The reason behind this is that it is *incredibly difficult* for a user to know where the default value for the SimObject (in this case the number of entries in the "cache") comes from if there are defaults in the command line parameters, SimObject file, and overridden in scripts.

With my new implementation, you only set the number of entries and nothing is "hard coded."
You can also have an associativity that is not fully associative and it should still work correctly.

mattsinc commented on src/arch/amdgpu/vega/pagetable_walker.hh:
```diff
@@ -193,12 +258,16 @@ class Walker : public ClockedObject
         deviceRequestorId(999), system(p.system)
     {
-        DPRINTF(GPUPTWalker, "Walker::Walker %p\n", this);
+      DPRINTF(GPUPTWalker, "Walker::Walker %p\n", this);
```
pretty sure indentation is wrong here?

TomXia replied:
Sorry, missed this. Fixed in the new commit.

mattsinc reviewed (changes requested):
There is one minor change with this code directly, but I'm honestly now sure how to review this anymore given the large changes separately on @powerjg 's branch that seem mostly orthogonal to the original goal of this patch ... as it seems to mostly be about refactoring the what the GPU PTW is working to follow a different format.  I'm not opposed to that, but I also am not sure squashing that and this together makes sense as this commit is about adding a PWC, whereas that commit is about using a different style for modeling the PWC.

I don't really know what to do anymore so if @powerjg is happy and the code actually works with all these changes (which I find difficult to follow as they are now scattered across at least two different places, in two different styles), then we should just merge ...

mattsinc commented:
@TomXia just checking if you have made at least the one minor change I highlighted?

TomXia committed:
arch-vega: Update PWC implemenatation

Add a new indexing policy cache that is specifically for a fully
associative TLB-like thing. Also move some code out of the page table
walker file into a new pwc file.

This change also fixes a spelling mistake entires->entries

TomXia commented:
Thanks.

I was indeed struggling passing the parameter to the indexing policy. Now everything looks better. Thanks again @powerjg  for making the change. I tested the implementation and fixed a small issue. Pushed in a new commit.

And Yes, the PWC is similar to TLB in some way. But it acts more like a cache and helps to reduce page table walk cost.

powerjg commented on src/arch/amdgpu/vega/page_walk_cache.hh:
```diff
@@ -0,0 +1,159 @@
+{
+
+// Page walk cache entry
+struct PWCEntry : public ReplaceableEntry
```
I just want to triple check, the intended implementation here is that each entry in the "page walk cache" is a *single* PTE. I.e., the page walk cache is *exactly the same* as a second level TLB except that it is tagged by the address of the PTE instead of the virtual address.

I have to say that I don't understand why the hardware would be that way, but if that's the way it is... 🤷

mattsinc replied:
Yes, each entry is a single PTE.  I don't think the PWC is the same as the L3 GPU TLB though (unless the L3 GPU TLB is the same number of entries) -- because the PWC is (I believe) only going to hold pages that recently missed in the L3 GPU TLB and are going to be sent there (so eventually they'll be cached again in the L3 GPU TLB, yes).  Conversely, the L3 GPU TLB would likely hold more pages if it is larger (and potentially evict them in a different order depending on reuse).  In that sense, I've been thinking of the PWC as being more like a victim buffer for the L3 GPU TLB, but maybe I'm wrong.

@abmerop I believe this is what you had recommended though, so you can chime in if I'm misrepresented anything.

abmerop replied:
PWC is caching intermediate nodes of the page table tree structure.  TLBs cache the leaf nodes.  The intention of the PWC is not to directly cache a translation but to reduce the number of round trips to memory (one for each level in the page table at most, possibly as low as 1 with PWC).

mattsinc replied:
I'll wait to make sure @powerjg is ok with @abmerop 's reply before merging

powerjg replied:
I think I understand the difference now. Thanks!
